Hw1

IAGrig opened the pull request.

IAGrig committed:
build: add database module and prettify initial configs

IAGrig committed:
feat: implement bot and all commands, add allowed domains

The bot starts with Spring, gets token from enviromental variables.
Command processing is based on Chain of Responsibility pattern.
The LinksHandler is responsible for processing and validating
incoming links for tracking.
The allowed domains are specified in the configuration files as
a single line separated by commas without spaces.

IAGrig committed:
feat: add database support

The Stub Database is a temporal in-memory database, which will be
replaced by a real database in the future.

IAGrig committed:
test: add tests for commands handling and LinksHandler

IAGrig committed:
feat: add a menu with commands to the bot and do a little refactoring

IAGrig committed:
fix: fix a bug in /list (links id were not displayed)

The user must to specify the link id to stop tracking the link using
/untrack, but these ids were not displayed in /list output.

IAGrig commented:
Стоит ли мне выделить отдельный класс, условно BotRunner, который будет реализовывать CommandLineRunner, создавать экземпляр бота и запускать его? Основную функциональность бота в это время отделить во второй класс.

damir2407 commented on bot/src/main/resources/application.yml:
```diff
@@ -1,5 +1,6 @@
 app:
   telegram-token: unset
```
Умеешь пользоваться переменными окружения? Тут можно написать что-то типа telegram-token: ${TELGRAM_TOKEN} и прокинуть значение в env и не бояться слить креды в репо. Хз какая у тебя ide, но вот как в интельке ультимейт: 
<img width="1466" alt="image" src="https://github.com/IAGrig/tinkoff-backend-2024/assets/70958074/711f89a8-3df5-4e8e-bd63-a72690711ed1">

damir2407 replied:
А, нашел профиль dev, отбой)))

damir2407 commented on bot/src/main/java/edu/java/bot/configuration/application-dev.yaml:
```diff
@@ -0,0 +1,2 @@
+app:
+  telegram_api_key: ${TELEGRAM_API_KEY}
```
Во-первых: го хранить конфу прилы в resources пакете? Это специально отведенный пакет для таких нужд.
Во-вторых: кажется, будто бы нет смысла выносить токен tg бота в отдельный дев профиль. Дев профиль это больше про различия конфы от Прода/Теста, но у тебя же ТГ бот в любом случае должен быть с токеном, поэтому эта настройка общая и должна быть в application.yml, а уже в зависимости от контура, на котором развернется бот, там будут разные токены в env

damir2407 commented on bot/src/main/java/edu/java/bot/bot/Bot.java:
```diff
@@ -0,0 +1,68 @@
+public class Bot implements CommandLineRunner {
+
+    private final Environment env;
+    private final LinksHandler linksHandler;
```
Не используешь linksHandler, можно убрать

damir2407 commented on bot/src/main/java/edu/java/bot/bot/Bot.java:
```diff
@@ -0,0 +1,68 @@
+@Log4j2
+public class Bot implements CommandLineRunner {
+
+    private final Environment env;
```
Окей, я вижу что со спрингом есть небольшие трудности, давай вопросы спринга тогда отложим на след дз, в рамках этого менять не нужно. НО, если считаешь, что хочешь поправить в этой - отпишись

IAGrig replied:
Здравствуйте, со спрингом трудности действительно имеются. Как-то сумбурно нахватался разных понятий и механизмов, а целостной картины и понимания вообще нет. 
Не хотелось бы откладывать необходимые правки. Думаю, лучше начать разбираться с самого начала. Но если вы считаете, что уместнее будет отложить до следующего ДЗ, то давайте сделаем так.

damir2407 replied:
Привет! Давай все же отложем ибо не хочу блочить тебе выполнение следующих дз, а то в рамках этой спринга и не просят

damir2407 commented on bot/src/main/java/edu/java/bot/bot/Bot.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Autowired
+    public Bot(Environment env, LinksHandler linksHandler) {
+        this.env = env;
```
env юзаешь только в конструкторе, из полей класса можно убрать

damir2407 commented on bot/src/main/java/edu/java/bot/bot/Bot.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Override
+    public void run(String... args) throws Exception {
```
Кому throws кинул, знаешь?)

Давай вообщем уберем его, тебе не подсвечивает такое?
<img width="869" alt="image" src="https://github.com/IAGrig/tinkoff-backend-2024/assets/70958074/8d454ec3-faea-46f3-b702-3d2fbc228a91">

damir2407 commented on bot/src/main/java/edu/java/bot/bot/Bot.java:
```diff
@@ -0,0 +1,68 @@
+    private void setUpdatesListener() {
+        bot.setUpdatesListener(updates -> {
+            updates.forEach(update -> {
+                log.info(update.toString());
```
А если ботом решилась воспользоваться не понимающая что происходит бабушка и отправила свой пароль? Будешь хранить в логах?)))

IAGrig replied:
Тут не до конца понимаю. Как лучше сделать: логировать информацию частично или вообще ничего не логировать? Если логировать частично, то что стоит оставить, а что убрать?

damir2407 replied:
Имхо лучше сделать лог по типу: "пришел пользователь с chatId = x, с командой = y"

damir2407 commented on bot/src/main/java/edu/java/bot/bot/commands/CommandsManager.java:
```diff
@@ -0,0 +1,33 @@
+import edu.java.bot.bot.links.LinksHandler;
+
+public class CommandsManager {
+    private final LinksHandler linksHandler;
```
юзаешь только в конструкторе

damir2407 commented on bot/src/main/java/edu/java/bot/bot/commands/CommandsManager.java:
```diff
@@ -0,0 +1,33 @@
+public class CommandsManager {
+    private final LinksHandler linksHandler;
+    private final CommandHandler firstHandler;
+    private final Database database;
```
юзаешь только в конструкторе

damir2407 commented on bot/src/main/java/edu/java/bot/bot/commands/HelpCommandHandler.java:
```diff
@@ -0,0 +1,25 @@
+    private final static String HELP_MESSAGE = """
+        Welcome to Link Updates Notifier. \n
+        Here is list of all available commands:
+        /help - show this message.
```
/start потерял

damir2407 commented on bot/src/main/java/edu/java/bot/bot/Bot.java:
```diff
@@ -0,0 +1,68 @@
+        BotCommand helpCommand = new BotCommand("/help", "show help message about all commands");
+        BotCommand listCommand = new BotCommand("/list", "show the links you are tracking");
+        BotCommand trackCommand = new BotCommand("/track", "start tracking the link");
+        BotCommand untrackCommand = new BotCommand("/untrack", "stop tracking the link");
```
/start потерял

damir2407 commented on bot/src/main/java/edu/java/bot/bot/commands/StartCommandHandler.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    public String executeCommand(Message message) {
+        return String.format(
+            "Welcome, %s. Try /help command to discover all functionality.",
```
> /start -- зарегистрировать пользователя

Это по заданию. У тебя же просто выводится имя пользователя и ты можешь бесконечно "регаться" по /start

damir2407 commented on bot/src/main/java/edu/java/bot/bot/links/LinksHandler.java:
```diff
@@ -0,0 +1,24 @@
+@Component
+@RequiredArgsConstructor
+public class LinksHandler {
+    private final static String BEFORE_DOMAIN_URL_PATTERN = "https?://(?:[a-z]+\\.)*";
```
Почитай про URI класс, там вроде просто скидываешь строку и смотришь запарсилось или нет, без танцев с бубном, удобная вещь)

damir2407 commented:
> Стоит ли мне выделить отдельный класс, условно BotRunner, который будет реализовывать CommandLineRunner, создавать экземпляр бота и запускать его? Основную функциональность бота в это время отделить во второй класс.


Не, тебе вообще не нужен CommandLineRunner, ты по факту даже ничего не запускаешь, а просто сетишь команды в меню + устанавливаешь апдейт лиснер. Это все ты можешь в конструкторе пока сделать. Сам бин создастся и будет готов принимать запросы с тг

damir2407 commented on bot/src/test/java/edu/java/bot/commands/CommandsManagerTest.java:
```diff
@@ -0,0 +1,39 @@
+    @DisplayName("CommandsManager correct test")
+    public void commandManagerCorrectTest() {
+        Database mockDatabase = Mockito.mock(Database.class);
+        LinksHandler mockLinksHandler = Mockito.mock(LinksHandler.class);
```
в обоих из методов создаешь mockDatabase, mockLinksHandler, manager. Вынеси выше в поля класса

damir2407 commented on bot/src/test/java/edu/java/bot/commands/HelpCommandHandlerTest.java:
```diff
@@ -0,0 +1,63 @@
+    @Test
+    @DisplayName("/help correct test")
+    public void helpCorrectTest() {
+        Database mockDatabase = Mockito.mock(Database.class);
```
также с выносом везде

IAGrig committed:
fix: correct all defects

The application configuration has been moved from the dev profile
to the general one. Unused fields in classes have been removed.
Fixed the processing of the /start command and the processing of
allowed domains, added functionality for tracking registered users.

IAGrig committed:
test: fix the tests - replace junit4 with junit5
